### DISCUSSION TITLE
Add a parameter hourstep

### DISCUSSION THREAD

**Discussion by gitbeyond** [apache/skywalking#8250] [Ideas]
# English version
I have a skywalking cluster that write the data to a elasticsearch cluster.

The elasticsearch cluster has 3 master nodes and 4 data nodes.

These index parameters of skywalking are:
```bash
SW_STORAGE_ES_INDEX_SHARDS_NUMBER=3
SW_STORAGE_ES_INDEX_REPLICAS_NUMBER=1
SW_STORAGE_ES_SUPER_DATASET_INDEX_REPLICAS_NUMBER=5
SW_STORAGE_ES_SUPER_DATASET_INDEX_SHARDS_FACTOR=1
```
Therefore, the shards of the pattern sw_segment-* is 15 and the replicas is 1.

This is not my ideal situation, because in this case, if a data node goes down, the cluster will become red, unable to write, and some reads will fail(The recovery process is very slow).

So I plan to set `SW_STORAGE_ES_SUPER_DATASET_INDEX_REPLICAS_NUMBER` to 1. So the shards of `sw_segment-*` will be 3. In this case, three shards and one replia. If a data node is damaged, the cluster is still normally writable.

However, a new problem has arisen, elasticsearch officially recommending that single-slice data not exceed 50G(https://www.elastic.co/guide/en/elasticsearch/reference/current/size-your-shards.html), but currently sw_segment have a daily data volume of 200G+. As a result, the average number of three fragments exceeds 50G per fragment, which will continue to increase.

Although it can be solved by adding machines, the current cluster does not have high storage utilization. Adding nodes is a bit wasteful.

Therefore, if the index of skywalking sw_segment can support the "hour" level, for example, by adding the hourstep parameter, it can be set to 10 hours or even 1 hour to solve this problem.

# 中文版本
我有一个skywalking集群，使用了elasticsearch作为存储集群。

这个elasticsearch集群，拥有3个master和4个data节点。

相关的索引参数设置情况为:
```bash
SW_STORAGE_ES_INDEX_SHARDS_NUMBER=3
SW_STORAGE_ES_INDEX_REPLICAS_NUMBER=1
SW_STORAGE_ES_SUPER_DATASET_INDEX_REPLICAS_NUMBER=5
SW_STORAGE_ES_SUPER_DATASET_INDEX_SHARDS_FACTOR=1
```
所以`sw_segment-*`索引的分片为15,副本为1。

那这不是我理想的情况，因为这种情况下，出现了一个data node宕机的情况，那么集群将成为red,无法写入，且部分读取会失效。

所以我计划将`SW_STORAGE_ES_SUPER_DATASET_INDEX_REPLICAS_NUMBER`设为1，让`sw_segment-*`的分片为`3`,这样的话，(实际上，这里需要修改index template)3分片加上1个副本，如果出现一个data node损坏的情况，那么集群至少仍然是正常可写的。

但是又出现了新的问题，es官方建议单分片数据不超过50G,但是目前的sw_segment一天的数据量在200G+, 这样一来，三个分片，平均每个分片就超过了50G了，后面此数据量还会继续增加。

虽然可以通过添加机器来解决，但是当前的集群，在储存的利用率上并不高。加节点有点浪费。

所以就想：如果skywalking的sw_segment的索引可以支持到“小时”级别就好了，比如增加hourStep参数，那设置为10个小时，甚至1个小时，就可以解决这个问题了。

**Comment by wu-sheng** [comment]
Adding this kind of option would be very painful, and causes a lot of changes in all relative mechanism.
I don't think this is good idea.

Also, making index created per N hours, which means you are going to face index creating in day time, this is also not recommended for a product env usage.

**Reply by gitbeyond** [reply]
> Adding this kind of option would be very painful, and causes a lot of changes in all relative mechanism. I don't think this is good idea.
> 
> Also, making index created per N hours, which means you are going to face index creating in day time, this is also not recommended for a product env usage.

Well, I'll think of another way. Thank you for your reply.
